adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java

```diff
@@ -183,66 +185,95 @@ private void handleEndpointConnection(final MqttEndpoint endpoint) {
-                                close(endpoint);
-                            } else {
-                                LOG.trace("successfully processed message [client ID: {}, topic: {}, QoS: {}]", endpoint.clientIdentifier(),
-                                        resource, message.qosLevel());
-                            }
-                        });
+    private Handler<MqttPublishMessage> createMqttPublishMessageHandler(final MqttEndpoint endpoint) {
+        return message -> {
 
-                        // check that MQTT client tries to publish on topic with device_id same as on connection
-                        if (resource.getResourceId().equals(endpoint.clientIdentifier())) {
+            LOG.trace("received message [client ID: {}, topic: {}, QoS: {}, payload {}]", endpoint.clientIdentifier(), message.topicName(),
+                    message.qosLevel(), message.payload().toString(Charset.defaultCharset()));
 
-                            Future<String> assertionTracker = getRegistrationAssertion(endpoint, resource);
-                            Future<MessageSender> senderTracker = getSenderTracker(message, resource);
+            try {
 
-                            CompositeFuture.all(assertionTracker, senderTracker).compose(ok -> {
-                                doUploadMessage(resource.getTenantId(), assertionTracker.result(), endpoint, message, senderTracker.result(), messageTracker);
-                            }, messageTracker);
+                final ResourceIdentifier resource = ResourceIdentifier.fromString(message.topicName());
 
+                // if MQTT client doesn't specify device_id then closing connection (MQTT has now way for errors)
+                if (resource.getResourceId() == null) {
+                    close(endpoint);
+                } else {
 
+                    Future<Void> messageTracker = Future.future();
+                    messageTracker.setHandler(s -> {
+                        if (s.failed()) {
+                            LOG.debug("cannot process message [client ID: {}, topic: {}, QoS: {}]: {}", endpoint.clientIdentifier(),
+                                    resource, message.qosLevel(), s.cause().getMessage());
+                            close(endpoint);
                         } else {
-                            // MQTT client is trying to publish on a different device_id used on connection (MQTT has no way for errors)
-                            messageTracker.fail("client not authorized");
+                            LOG.trace("successfully processed message [client ID: {}, topic: {}, QoS: {}]", endpoint.clientIdentifier(),
+                                    resource, message.qosLevel());
                         }
+                    });
 
-                    }
+                    // check that MQTT client tries to publish on topic with device_id same as on connection
+                    if (resource.getResourceId().equals(endpoint.clientIdentifier())) {
```

Comment:
according to the proposal made in #86 this check is not necessary anymore, is it?